### PR TITLE
Pausable

### PR DESCRIPTION
**Pull request by cryptoAtwill**
Add `pausable` to subnet actor such that operations with `whenNotPaused` will be blocked when necessary.

### CODE REVIEW

File: src/lib/LibPausable.sol [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    error AlreadyPaused();
+
+    modifier whenNotPaused() {
+        PausableStorage storage s = pausableStorage();
+        if (s.paused) revert AlreadyPaused();
+        _;
+    }
+
+    /// @notice sets if to pause the contract
+    function isPausabled() external view returns(bool) {
```
Comment (dnkolegov) [inline src/lib/LibPausable.sol:22]: I think it should be named `isPaused`

Comment (dnkolegov) [reply src/lib/LibPausable.sol:22]: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Pausable.sol

File: src/lib/LibPausable.sol [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        PausableStorage storage s = pausableStorage();
+        return s.paused;
+    }
+
+    /// @notice sets if to pause the contract
+    function setPausable(bool paused) internal {
```
Comment (dnkolegov) [inline src/lib/LibPausable.sol:28]: Why do you need an argument? Why do we need setPausable(false)?
It should be named like `pause()`.
I suggest using functions names and semantics from https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Pausable.sol

File: src/subnet/SubnetActorManagerFacet.sol [hunk 3]
```diff
@@ -238,7 +252,7 @@ contract SubnetActorManagerFacet is ISubnetActor, SubnetActorModifiers, Reentran
     /// @notice method that allows a validator to leave the subnet
     /// @dev it also return the validators initial balance if the
     /// subnet was not yet bootstrapped.
-    function leave() external nonReentrant notKilled {
+    function leave() external whenNotPaused nonReentrant notKilled {
```
Comment (dnkolegov) [inline src/subnet/SubnetActorManagerFacet.sol:255]: `nonreentrant` always should be first one

File: src/subnet/SubnetActorManagerFacet.sol [hunk 4]
```diff
@@ -286,17 +300,17 @@ contract SubnetActorManagerFacet is ISubnetActor, SubnetActorModifiers, Reentran
     }
 
     /// @notice Validator claims their released collateral
-    function claim() external nonReentrant {
+    function claim() external whenNotPaused nonReentrant {
```
Comment (dnkolegov) [inline src/subnet/SubnetActorManagerFacet.sol:303]: nonreentrant always should be the first one

File: src/subnet/SubnetActorManagerFacet.sol [hunk 5]
```diff
@@ -286,17 +300,17 @@ contract SubnetActorManagerFacet is ISubnetActor, SubnetActorModifiers, Reentran
...
         LibStaking.claimCollateral(msg.sender);
     }
 
     /// @notice Relayer claims its reward
-    function claimRewardForRelayer() external nonReentrant {
+    function claimRewardForRelayer() external whenNotPaused nonReentrant {
```
Comment (dnkolegov) [inline src/subnet/SubnetActorManagerFacet.sol:308]: nonreentrant always should be the first one

**Review by dnkolegov — CHANGES_REQUESTED** [review]
I suggest using https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Pausable.sol as a reference: use functions names and semantics.

**Comment by dnkolegov** [comment]
> Add `pausable` to subnet actor such that operations with `whenNotPaused` will be blocked when necessary.

Are we going to add pausability for Gateways?

**Comment by cryptoAtwill** [comment]
close temporarily to avoid generating `out` and `binding`.

**Comment by cryptoAtwill** [comment]
@dnkolegov I have updated the code based on your feedbacks. Please take another look.

**Review by dnkolegov — APPROVED** [review]

**Pull request merged**